File: lib/services/executors/claude-agent/src/cli-stream-parser.test.ts
// Copyright © 2026 Fall Guy Consulting.
// Dual-licensed under AGPL-3.0-or-later or a Fall Guy Consulting commercial
// license. See LICENSE.agpl and COPYRIGHT at the repo root.

import { describe, it, expect } from "vitest";
import { makeCliStreamParser } from "./cli-stream-parser.js";

describe("makeCliStreamParser", () => {
  it("emits tool_use_start on assistant message with tool_use content block", () => {
    const p = makeCliStreamParser();
    const evs = p.push(
      JSON.stringify({
        type: "assistant",
        message: {
          content: [{ type: "tool_use", id: "toolu_1", name: "Bash", input: {} }],
        },
      }) + "\n",
    );
    expect(evs).toEqual([{ kind: "tool_use_start", id: "toolu_1", name: "Bash" }]);
  });

  it("emits tool_use_end on user message with tool_result content block", () => {
    const p = makeCliStreamParser();
    const evs = p.push(
      JSON.stringify({
        type: "user",
        message: {
          content: [{ type: "tool_result", tool_use_id: "toolu_1", content: "ok" }],
        },
      }) + "\n",
    );
    expect(evs).toEqual([{ kind: "tool_use_end", id: "toolu_1" }]);
  });

  it("buffers partial lines across chunks", () => {
    const p = makeCliStreamParser();
    const line = JSON.stringify({
      type: "assistant",
      message: { content: [{ type: "tool_use", id: "toolu_2", name: "Read" }] },
    });
    const halves = [line.slice(0, 30), line.slice(30) + "\n"];
    const a = p.push(halves[0]);
    expect(a).toEqual([]);
    const b = p.push(halves[1]);
    expect(b).toEqual([{ kind: "tool_use_start", id: "toolu_2", name: "Read" }]);
  });

  it("ignores non-JSON lines and shapes we don't care about", () => {
    const p = makeCliStreamParser();
    const evs = p.push(
      "not json\n" +
        JSON.stringify({ type: "system", subtype: "init" }) +
        "\n" +
        JSON.stringify({ type: "result", subtype: "success" }) +
        "\n",
    );
    expect(evs).toEqual([]);
  });

  it("emits multiple tool_use events from a single assistant message with parallel calls", () => {
    const p = makeCliStreamParser();
    const evs = p.push(
      JSON.stringify({
        type: "assistant",
        message: {
          content: [
            { type: "text", text: "thinking" },
            { type: "tool_use", id: "toolu_A", name: "Bash" },
            { type: "tool_use", id: "toolu_B", name: "Read" },
          ],
        },
      }) + "\n",
    );
    expect(evs).toEqual([
      { kind: "tool_use_start", id: "toolu_A", name: "Bash" },
      { kind: "tool_use_start", id: "toolu_B", name: "Read" },
    ]);
  });
});

File: lib/services/executors/claude-agent/clistreamparser_test.go
// Copyright © 2026 Fall Guy Consulting.
// Dual-licensed under AGPL-3.0-or-later or a Fall Guy Consulting commercial
// license. See LICENSE.agpl and COPYRIGHT at the repo root.

package claudeagent

import (
	"reflect"
	"strings"
	"testing"
)

func TestStreamParserEmitsToolUseStart(t *testing.T) {
	p := NewCliStreamParser()
	evs := p.Push(`{"type":"assistant","message":{"content":[{"type":"tool_use","id":"toolu_1","name":"Bash","input":{}}]}}` + "\n")
	want := []CliProgressEvent{{Kind: "tool_use_start", ID: "toolu_1", Name: "Bash"}}
	if !reflect.DeepEqual(evs, want) {
		t.Fatalf("events = %+v, want %+v", evs, want)
	}
}

func TestStreamParserEmitsToolUseEnd(t *testing.T) {
	p := NewCliStreamParser()
	evs := p.Push(`{"type":"user","message":{"content":[{"type":"tool_result","tool_use_id":"toolu_1","content":"ok"}]}}` + "\n")
	want := []CliProgressEvent{{Kind: "tool_use_end", ID: "toolu_1"}}
	if !reflect.DeepEqual(evs, want) {
		t.Fatalf("events = %+v, want %+v", evs, want)
	}
}

func TestStreamParserBuffersPartialLines(t *testing.T) {
	p := NewCliStreamParser()
	line := `{"type":"assistant","message":{"content":[{"type":"tool_use","id":"toolu_2","name":"Read"}]}}`
	if evs := p.Push(line[:30]); len(evs) != 0 {
		t.Fatalf("expected no events for partial line, got %+v", evs)
	}
	evs := p.Push(line[30:] + "\n")
	want := []CliProgressEvent{{Kind: "tool_use_start", ID: "toolu_2", Name: "Read"}}
	if !reflect.DeepEqual(evs, want) {
		t.Fatalf("events = %+v, want %+v", evs, want)
	}
}

func TestStreamParserIgnoresNonJSONAndOtherShapes(t *testing.T) {
	p := NewCliStreamParser()
	evs := p.Push("not json\n" +
		`{"type":"system","subtype":"init"}` + "\n" +
		`{"type":"result","subtype":"success"}` + "\n")
	if len(evs) != 0 {
		t.Fatalf("expected no events, got %+v", evs)
	}
}

func TestStreamParserCapsUnboundedPendingLine(t *testing.T) {
	p := NewCliStreamParser()
	chunk := strings.Repeat("x", 1024*1024)
	for i := 0; i < 16; i++ {
		if evs := p.Push(chunk); len(evs) != 0 {
			t.Fatalf("expected no events for a newline-less oversized chunk, got %+v", evs)
		}
	}
	if got := p.buf.Len(); got > maxPendingStdoutLineBytes {
		t.Fatalf("pending buffer = %d bytes, want capped at %d (an adversarial newline-less stream must not grow memory unbounded)", got, maxPendingStdoutLineBytes)
	}
}

func TestStreamParserEmitsMultipleParallelToolUses(t *testing.T) {
	p := NewCliStreamParser()
	evs := p.Push(`{"type":"assistant","message":{"content":[` +
		`{"type":"text","text":"thinking"},` +
		`{"type":"tool_use","id":"toolu_A","name":"Bash"},` +
		`{"type":"tool_use","id":"toolu_B","name":"Read"}]}}` + "\n")
	want := []CliProgressEvent{
		{Kind: "tool_use_start", ID: "toolu_A", Name: "Bash"},
		{Kind: "tool_use_start", ID: "toolu_B", Name: "Read"},
	}
	if !reflect.DeepEqual(evs, want) {
		t.Fatalf("events = %+v, want %+v", evs, want)
	}
}
